### PR TITLE
Iterate over py::list objects by value instead of by reference.

### PR DESCRIPTION
**Pull request by Ukalnins**
`py::list` iterator has two implementations depending if pybind11 is compiled for cpython or pypy. For cpython the iterator returns essentially `PyObject*`, so taking a reference is not needed. For pypy, returned object is a temporary, so it requires that iteration happens by value instead of reference. Currently it's the only problem for building duckdb for pypy.

### CODE REVIEW

**Comment by Ukalnins** [comment]
Hey.  Even though duckdb does not officially support pypy, it seemed from previous discussions, that the team is open for it to work with pypy. I hope that this is something that can be merged as it does not really impact cpython support and is reasonably simple fix.

**Review by Tishj — APPROVED** [review]
Thanks, if CI is happy then I'm happy

Just for clarification though, when we iterate through a py::list, I assume it makes `py::object`s which require the GIL as it invokes a refcount++ and refcount-- in the constructor and destructor respectively.

I don't really know what it was doing before, I imagine it was still creating `py::object &` ?

Im assuming that it's `py::object` and not `py::handle` because indexing into a py::list or other iterables produce a `pybind11::detail::accessor`, which when using `auto` turns into a `py::object`

-------------------------------------

So it was a refcounting issue?
Basically returning a reference to an object that lives on the stack instead of returning by value?

**Comment by Ukalnins** [comment]
As I mentioned, the 
```
auto items = py::list(..);
for(auto item: items) { 
  ...
}
```
has different behaviour in pybind for cpython and pypy. Hence the compile error.

As far as I understand, even iterating by value won't increase the refcount for cpython. But it will for pypy.

For cpython what happens is that `py::list` calls `PySequence_Fast_ITEMS` to return the `PyObject**` of the underlying list and iteration in the loop runs over `const py::handle` objects that are just wrappers around `PyObject*` (without any ref counting).

For pypy what happens is that `py::list::begin()` constructs a iterator object that calls `PySequence_GetItem(obj.ptr(), index)` for each element to construct `py::object`.

The end result, as far as I understand, is that cpython will not suffer any performance degradation from this, but pypy will be slower than cpython.

**Comment by Ukalnins** [comment]
Started digging into pybind code and got off track a bit.

The problem is that the iterators return reference objects by value. For cpython the returned object is `const handle`, which can be used as `const handle&`, but for pypy it returns `object` which turns to `object&` and you get error about inability to bind the returned value to non-const reference.

**Comment by Ukalnins** [comment]
As I wrote the last comment, I understood, that we can just actually rewrite the loops as:
```
for (const auto& item: items) { ... }
```
and it would work the same. In the end it would generate the same underlying code, but maybe it would look better. 

Let me know if you want for me to update  the PR with `const&`, I'm not sure which better matches duckdb style guides and general approach.
